# Project 10 -- Niharika Rai

**TA Help:** (for instance) John Smith, Alice Jones, etc., list names of any TAs who helped you

- For example: Help with figuring out how to write a function (describe the tasks that they helped you with)

**Collaboration:** My Friend in CS, My Uncle, Another Student, etc., list names of any other people who helped you

(describe the tasks that they helped you with)
- For example: helped figuring out how to load the dataset.
- Another example: helped debug error with my plot.

**Internet Resources:** Stack Exchange, Stack Overflow, etc.

(describe any information that you learned from internet resources, including the URLs)
- data frames in Pandas versus R from StackOverflow  https://stackoverflow.com/questions/8991709/why-were-pandas-merges-in-python-faster-than-data-table-merges-in-r-in-2012

**ChatGPT, Gemini, Claude, etc:** Any language models or generative AI chatbots that helped you.

(if you used any such tools, please tell us here)
- For example:  I asked ChatGPT how to define a new data frames
- Another example:  Gemini told me how to make a function for sorting my data

- ***Link to AI Chat History***: None: I did not use any AI tool to complete this project. Please share a link to your chat if you used AI (ex. ChatGPT Shared Links)
**OVERALL MESSAGE:** Any time that you used anything except your brain to solve the questions in these projects, you need to disclose such resources at the start of the project, with details about your usage of the tools.

**YOUR OWN WORK:** Even when you utilize other resources, do NOT just copy and paste.  Write all explanations in your own words, using several sentences in English, which are understandable and which you wrote (and did not just copy and paste).

## Question 1

In [4]:
from pymongo import MongoClient
import pandas as pd
import json

# create a client to connect to the local MongoDB instance
client = MongoClient('mongodb://localhost:27017/')

# Create or connect to a database
db = client['movies_db']

# Create or connect to a collection
movies = db['movies']

In [33]:
# Read the CSV file and convert to MongoDB documents
df = pd.read_csv('/anvil/projects/tdm/data/MovieLens/movies_metadata.csv')

# Convert DataFrame to list of dictionaries
movies_data = df.to_dict('records')

# Insert the data into MongoDB
result = movies.insert_many(movies_data)
print(f"Inserted {len(result.inserted_ids)} movies")

/tmp/ipykernel_307/3709195004.py:2: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/anvil/projects/tdm/data/MovieLens/movies_metadata.csv')


Inserted 45466 movies


In [6]:
# Find all movies (limit to first 5 for display)
sample_movies = list(movies.find().limit(5))
print(f"Total movies in collection: {movies.count_documents({})}")
print("\nSample movie documents:")
for movie in sample_movies:
    print(f"Title: {movie.get('title', 'N/A')}")
    print(f"Release Date: {movie.get('release_date', 'N/A')}")
    print(f"Vote Average: {movie.get('vote_average', 'N/A')}")
    print("---")

Total movies in collection: 45466

Sample movie documents:
Title: Toy Story
Release Date: 1995-10-30
Vote Average: 7.7
---
Title: Jumanji
Release Date: 1995-12-15
Vote Average: 6.9
---
Title: Grumpier Old Men
Release Date: 1995-12-22
Vote Average: 6.5
---
Title: Waiting to Exhale
Release Date: 1995-12-22
Vote Average: 6.1
---
Title: Father of the Bride Part II
Release Date: 1995-02-10
Vote Average: 5.7
---


In [7]:
# Get the structure of a movie document
sample_movie = movies.find_one()
print("Movie document structure:")
for key, value in sample_movie.items():
    print(f"{key}: {type(value).__name__}")
    if isinstance(value, str) and len(str(value)) > 100:
        print(f"  Value: {str(value)[:100]}...")
    else:
        print(f"  Value: {value}")
    print()

Movie document structure:
_id: ObjectId
  Value: 69082f100d14805a4cf42cbe

adult: str
  Value: False

belongs_to_collection: str
  Value: {'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'ba...

budget: str
  Value: 30000000

genres: str
  Value: [{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]

homepage: str
  Value: http://toystory.disney.com/toy-story

id: str
  Value: 862

imdb_id: str
  Value: tt0114709

original_language: str
  Value: en

original_title: str
  Value: Toy Story

overview: str
  Value: Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto ...

popularity: float
  Value: 21.946943

poster_path: str
  Value: /rhIRbceoE9lR4veEXuwCC2wARtG.jpg

production_companies: str
  Value: [{'name': 'Pixar Animation Studios', 'id': 3}]

production_countries: str
  Value: [{'iso_3166_1': 'US', 'name': 'United States of America'}]

release_dat

1.1 I used the sample query in the question to achieve this. 
1.2 I used the sample query in the question to achieve this.
1.3 I used the sample query in the question to achieve this.
1.4 Based on the outputs, there are 45466 movies in the collection. The structure of the movie documents shows that it mostly contains columns/fields that are strings, float, bool and ObjectId. The different fields available are budget, genres, _id, id, adult, title, video, vote_average, vote_count, tagline, status, spoken_languages, runtime, revenue, release_date, popularity, belongs_to_collection, homepage, imdb_id, original_language, original_title, poster_path, overview, production_companies and production_countries. The sample movie titles are Toy Story, Jumanji, Grumpier Old Men, Waiting to Exhale, and Father of the Bride Part II.    

## Question 2

In [8]:
# Find movies with rating above 8.0
high_rated_movies = list(movies.find({"vote_average": {"$gt": 8.0}}))
print(f"Movies with rating above 8.0: {len(high_rated_movies)}")
for movie in high_rated_movies[:5]:  # Show first 5
    print(f"{movie['title']} - Rating: {movie['vote_average']}")

Movies with rating above 8.0: 1130
Se7en - Rating: 8.1
The Usual Suspects - Rating: 8.1
Taxi Driver - Rating: 8.1
Reckless - Rating: 10.0
Star Wars - Rating: 8.1


In [11]:
recent_good_movies = list(movies.find({
    "$and": [
        {"vote_average": {"$gt": 6.0}},
        {"release_date": {"$regex": "199[0-9]"}}]
}))
print(f"Good movies released after 1990: {len(recent_good_movies)}")
for movie in recent_good_movies[:5]:  # Show first 5
    print(f"{movie['title']} - Rating: {movie['vote_average']}, Year: {movie['release_date'][:4]}")

Good movies released after 1990: 2578
Toy Story - Rating: 7.7, Year: 1995
Jumanji - Rating: 6.9, Year: 1995
Grumpier Old Men - Rating: 6.5, Year: 1995
Waiting to Exhale - Rating: 6.1, Year: 1995
Heat - Rating: 7.7, Year: 1995


In [13]:
comedy_movies = list(movies.find({
    "genres": {"$regex": "Comedy", "$options": "i"}  # Case insensitive search
}))
print(f"Comedy movies: {len(comedy_movies)}")
for movie in comedy_movies[:5]:  # Show first 5
    print(f"{movie['title']} - Genres: {movie.get('genres', 'N/A')}")

Comedy movies: 13182
Toy Story - Genres: [{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
Grumpier Old Men - Genres: [{'id': 10749, 'name': 'Romance'}, {'id': 35, 'name': 'Comedy'}]
Waiting to Exhale - Genres: [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
Father of the Bride Part II - Genres: [{'id': 35, 'name': 'Comedy'}]
Sabrina - Genres: [{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]


2.1 I used the sample query in the question to achieve this.
2.2 I used the sample query in the question and modified it based on the hint to achieve this.
2.3 I used the sample query in the question and modified it based on the hint to achieve this.
2.4 Based on the outputs, the number of movies with rating above 6.0 and released in the 1990s were 2578. The number of comedy movies were 13182.  

## Question 3

In [14]:
# Find all movies sorted by rating (descending)
movies_by_rating = list(movies.find().sort("vote_average", -1))
print("Movies sorted by rating (highest first):")
for movie in movies_by_rating[:5]:  # Show top 5
    print(f"{movie['title']} - Rating: {movie['vote_average']}")

Movies sorted by rating (highest first):
Reckless - Rating: 10.0
Girl in the Cadillac - Rating: 10.0
The Haunted World of Edward D. Wood, Jr. - Rating: 10.0
Carmen Miranda: Bananas Is My Business - Rating: 10.0
Other Voices Other Rooms - Rating: 10.0


In [15]:
# Find the highest rated movie
highest_rated = movies.find().sort("vote_average", -1).limit(1)
top_movie = list(highest_rated)[0]
print(f"Highest rated movie: {top_movie['title']} - Rating: {top_movie['vote_average']}")

Highest rated movie: Reckless - Rating: 10.0


In [23]:
# Find the 3 most popular movies released after 2000
highest_popularity = movies.find({"release_date": {"$regex": "200[0-9]|201[0-9]|202[0-9]"}}).sort("popularity", -1).limit(3)
top_3_movies = list(highest_popularity)
for movie in top_3_movies:
    print(f"{movie['title']} - Popularity: {movie['popularity']}")

The Devil's Candy - Popularity: 9.968649
The Marine 5: Battleground - Popularity: 9.930139
Me casé con un boludo - Popularity: 9.915094


In [24]:
total_movies = movies.count_documents({})
high_rated_count = movies.count_documents({"vote_average": {"$gt": 7.0}})
recent_movies_count = movies.count_documents({"release_date": {"$regex": "199[0-9]"}})

print(f"Total movies: {total_movies}")
print(f"Movies with rating above 7.0: {high_rated_count}")
print(f"Movies released in 1990s: {recent_movies_count}")

Total movies: 45466
Movies with rating above 7.0: 7714
Movies released in 1990s: 5677


3.1 I ran the sample queries for sorting and limiting with the results shown above. 
3.2 I used the sample query from the question but modified it to get the required output. 
3.3 I used the sample query from the question but modified it to get the required output.
3.4 See results of all the queries executed above. 

## Question 4

In [25]:
# Count movies by genre
pipeline = [
    {"$group": {"_id": "$genres", "count": {"$sum": 1}}},
    {"$sort": {"count": -1}},
    {"$limit": 10}
]

genre_counts = list(movies.aggregate(pipeline))

print("Movies by genre (top 10):")
for genre in genre_counts:
    print(f"{genre['_id']}: {genre['count']} movies")

Movies by genre (top 10):
[{'id': 18, 'name': 'Drama'}]: 5000 movies
[{'id': 35, 'name': 'Comedy'}]: 3621 movies
[{'id': 99, 'name': 'Documentary'}]: 2723 movies
[]: 2442 movies
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]: 1301 movies
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]: 1135 movies
[{'id': 27, 'name': 'Horror'}]: 974 movies
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]: 930 movies
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]: 593 movies
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]: 532 movies


In [26]:
# Average rating by release year
pipeline = [
    {"$addFields": {
        "year": {"$substr": ["$release_date", 0, 4]}  # Extract year from date
    }},
    {"$match": {"year": {"$ne": ""}}},  # Filter out empty years
    {"$group": {
        "_id": "$year",
        "avg_rating": {"$avg": "$vote_average"},
        "count": {"$sum": 1}
    }},
    {"$sort": {"_id": -1}},  # Sort by year descending
    {"$limit": 10}
]

avg_ratings = list(movies.aggregate(pipeline))
print("Average rating by year (recent years):")
for year in avg_ratings:
    print(f"{year['_id']}: {year['avg_rating']:.2f} rating ({year['count']} movies)")

Average rating by year (recent years):
nan: nan rating (87 movies)
22: nan rating (1 movies)
2020: 0.00 rating (1 movies)
2018: 0.00 rating (5 movies)
2017: 5.86 rating (532 movies)
2016: 5.82 rating (1604 movies)
2015: 5.76 rating (1905 movies)
2014: 5.65 rating (1974 movies)
2013: 5.76 rating (1889 movies)
2012: 5.64 rating (1722 movies)


In [27]:
# Average rating by popularity ranges
pipeline = [
    {"$match": {
        "popularity": {"$exists": True, "$gt": 0},
        "vote_average": {"$gt": 0}
    }},
    {"$addFields": {
        "popularity_range": {
            "$switch": {
                "branches": [
                    {"case": {"$lt": ["$popularity", 10]}, "then": "Low (0-10)"},
                    {"case": {"$lt": ["$popularity", 50]}, "then": "Medium (10-50)"},
                    {"case": {"$lt": ["$popularity", 100]}, "then": "High (50-100)"}
                ],
                "default": "Very High (100+)"
            }
        }
    }},
    {"$group": {
        "_id": "$popularity_range",
        "avg_rating": {"$avg": "$vote_average"},
        "avg_popularity": {"$avg": "$popularity"},
        "count": {"$sum": 1}
    }},
    {"$sort": {"avg_popularity": 1}}
]

popularity_ratings = list(movies.aggregate(pipeline))
print("Average rating by popularity ranges:")
for pop_range in popularity_ratings:
    print(f"{pop_range['_id']}: {pop_range['avg_rating']:.2f} avg rating, {pop_range['avg_popularity']:.1f} avg popularity ({pop_range['count']} movies)")

Average rating by popularity ranges:
Low (0-10): 6.01 avg rating, 2.5 avg popularity (28464 movies)
Medium (10-50): 6.46 avg rating, 13.6 avg popularity (2653 movies)
High (50-100): 7.83 avg rating, 71.4 avg popularity (8 movies)
Very High (100+): 7.35 avg rating, 195.8 avg popularity (12 movies)


4.1 I used the sample query in the question to achieve this. 
4.2 I used the sample query in the question to achieve this. 
4.3 I used the sample query in the question to achieve this. This code feels a little long and complicated. 
4.4 Based on the outputs, most of the common genres were Drama, Comedy, Documentary and Horror. The average rating for movies in recent years has been a little over 5 but it was 0 for 2018 and 2020. The average ratings for the movies increased as their popularity ranges increased. It means that the more popular movies were rated higher as more people liked them. 

## Question 5

In [28]:
# Update a single document (find a specific movie by title)
result = movies.update_one(
    {"title": "Toy Story"},
    {"$set": {"our_rating": 9.5}}
)
print(f"Updated {result.modified_count} document")

# Check the update
updated_movie = movies.find_one({"title": "Toy Story"})
print(f"Toy Story's our rating: {updated_movie.get('our_rating', 'No rating')}")

Updated 1 document
Toy Story's our rating: 9.5


In [29]:
# Update all movies to add a 'watched' field
result = movies.update_many(
    {},
    {"$set": {"watched": False}}
)
print(f"Updated {result.modified_count} documents with watched field")

Updated 45466 documents with watched field


In [30]:
# Delete movies with very low ratings (be careful with this!)
result = movies.delete_many({"vote_average": {"$lt": 2.0}})
print(f"Deleted {result.deleted_count} movies with rating below 2.0")

# Check how many movies remain
remaining_movies = movies.count_documents({})
print(f"Remaining movies: {remaining_movies}")

Deleted 3206 movies with rating below 2.0
Remaining movies: 42260


In [31]:
# Delete movies from a specific year (example: 1990)
result = movies.delete_many({"release_date": {"$regex": "1990"}})
print(f"Deleted {result.deleted_count} movies from 1990")

# Check remaining movies
remaining_movies = movies.count_documents({})
print(f"Remaining movies: {remaining_movies}")

Deleted 394 movies from 1990
Remaining movies: 41866


In [32]:
# Show some sample remaining movies
sample_movies = list(movies.find().limit(3))
print("Sample remaining movies:")
for movie in sample_movies:
    print(f"- {movie['title']} ({movie.get('release_date', 'N/A')})")

Sample remaining movies:
- Toy Story (1995-10-30)
- Jumanji (1995-12-15)
- Grumpier Old Men (1995-12-22)


5.1 I used the sample query in the question to achieve this. 
5.2 I used the sample query in the question to achieve this. 
5.3 I used the sample query in the question to achieve this. 
5.4 I used the sample query in the question to achieve this. 
5.5 I used the sample query in the question to achieve this.  

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
